### PR TITLE
Feature/gh 11819 quick order e2e

### LINKED ISSUE #11819: E2E tests for Quick Order
Minimum test cases:

#### Adding and listing SKUs

Add SKU using form > [repeat multiple times] > test listing in quick order listing page

#### Add to cart

Add SKU using form > [repeat multiple times] > Add quick order to cart > Go to cart > verify added SKUs

### PR DESCRIPTION
**Pull request by dydome**
close #11819

### CODE REVIEW

File: projects/storefrontapp-e2e-cypress/cypress/sample-data/b2b-quick-order.ts
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/sample-data/b2b-quick-order.ts]: there already exist b2b products that you can use, which can be found under `projects/storefrontapp-e2e-cypress/cypress/sample-data/b2b-checkout.ts`

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts]: instead of hardcoding the `wait` in which will most likely fail one day, you can 

* Cypress is very fast and one main reason that we have flaky tests is that we usually forget to wait for the response for a set of data that we're asserting to.

```
- create an intercept, and intercept the exact endpoint that will get triggered when you click add to cart
- wait for the response to give you a status of 200.

example: with saved cart, I intercepted the request that gets triggered when I click save a cart from the cart page

export function interceptSaveCartEndpoint(cartCode: string) {
  cy.intercept(
    'PATCH',
    `${Cypress.env('OCC_PREFIX')}/${Cypress.env(
      'BASE_SITE'
    )}/users/current/carts/${cartCode}/save?*`
  ).as(SAVE_CART_ENDPOINT_ALIAS);

  return SAVE_CART_ENDPOINT_ALIAS;
}

----
how to use would be like 

1 - const variableName = interceptSaveCartEndpoint('someCartNumber')

2 - cy.get('save cart button').click();

3 - cy.wait(`@${alias}`).its('response.statusCode').should('eq', 200);

^The above helps make sure the we got the response back from what we requested.


```

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts]: It fails here ootb. Basically, it's telling you that your feature is not accessible. Please make sure it is before e2e too.

Screen Shot 2021-07-13 at 11 52 43 PM

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts]: Responding to the DM question, 

- reason why your waitForPage is not working locally is because your cypress.json is pointing to `https://40.76.109.9:9002` and your environment.ts is pointing to `dev3`. These values should match because the waitForPage function is trying to wait for a request from host 40.76.xxx, but your actual request is dev3.xxx, therefore not being able to intercept the proper requests that are firing.

Comment (bgambocjaviniar) [reply projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts]: To fix this issue, you can copy and paste this env obj and replace the existing env obj from your `cypress.json` file. Also, take note of the changes on `API_URL`, `BASE_SITE`, and OCC..._USER_ENDPOINT`

```
  "env": {
    "CLIENT_ID": "mobile_android",
    "CLIENT_SECRET": "secret",
    "API_URL": "https://spartacus-dev3.eastus.cloudapp.azure.com:9002",
    "BASE_SITE": "powertools-spa",
    "BASE_LANG": "en",
    "BASE_CURRENCY": "USD",
    "OCC_PREFIX": "/occ/v2",
    "OCC_PREFIX_USER_ENDPOINT": "orgUsers",
    "OCC_PREFIX_ORDER_ENDPOINT": "orders"
  }
```

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts:135]: I believe not all your test below is checking for accessibility - keyboarding.

GC would like the e2e test case for accessibility to be run at the last of the sequence for the e2e test.

```
describe('Quick order page', () => {})

describe('Cart page', () => {})

describe('Accessibility - keyboarding', () => {})
```

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]
1 - make sure your epic is perfect (green pipeline) before having these e2e as I'm noticing strange failures that might affect your e2e

2 - do not include these template, css, and non-relevant e2e things in this PR. Please create an issue and tackle those problems in another as you still have this epic and have room to improve. **I'd like this PR to only contain e2e relevant things as I'm assuming these changes are done to accommodate your e2e, which is not the right away to write e2e tests. You shouldn't be adding more things to your template, services, css, etc to make your e2e run. It should be done before writing the e2e** 

3 - will this only be a b2b feature, or it works for b2c too?

4 - drop instances of cy.wait(1000). Hardcoded waits shouldn't be used if possible as I notice in your e2es, they can be avoided by intercepting the correct requests and waiting for it to finish processing and getting a response back.

File: projects/storefrontapp-e2e-cypress/cypress.json
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress.json:12]: Before merging, don't forget to revert these changes in this file

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts]: in that same method, you can wait for this request to happen when adding a product to your quick order list instead of that hard-coded wait for 1s 
https://github.com/SAP/spartacus/pull/13083#discussion_r669258584

Screen Shot 2021-07-16 at 12 33 37 AM

Comment (bgambocjaviniar) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts]: Once you add this, it will fix the flakiness, where when it clears, they won't have a product coming out of nowhere to fail the test.

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts]: change this to beforeEach, and you can skip adding `quickOrder.clearList` in line 21, 30,40

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts]: is this still being used?

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts:89]: it would be great to add after this line to assert the global error message being thrown that it's being cleared

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]
Similar to before,

1 - are you sure the feature is working? Unit tests are failing. I found that when you add 3 same products, it will add 3 entries. But when you refresh, the first entry will have 3 quantity and still have the 2 other entries. Make sure the feature is functional + no blockers for 100% confidence in the feature. I ran the tests and there's 3 failing tests related to quick order similar to what the travis job is showing.

2 - Make sure your prettier is working which can be found at the bottom right of your vscode. You should see checkmarks beside the name because currently, it's saying one of your files isn't prettified. You can perform the package script `yarn prettier:fix` do solve this issue as well.

3 - Ask GC to redeploy the backend with your updated sample data for the CI as this feature is almost ready.

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts]: is this still being used?

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts
Comment (dydome) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts]: I cannot do it as for other tests in this 'describe' I cannot call clear list it will crash

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts
Comment (bgambocjaviniar) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts]: Question at hand again is about why you need to keep the `cy.wait(1000)`

But the screenshot above explains that when you add a product to that quick order form, you are loading that exact product to your list which fires that request.

Specifically, if you look again at the screenshot below, when the test enters the product code + enter action, we can see the xhr request being made. What I'm asking is every time you 'add' a product to that list, meaning that you load it, you wait for that request to finish happening.

Example:

```
export function addProductToTheList(productCode: string) {
  #### registerIntecerpt() ---> the `GET` request which is basically 'loading' the product so that you can populate your entries

  cy.get('.quick-order-form-input input').type(`${productCode}{enter}`);

##### cy.wait(thatInterceptAlias).its('response.statusCode').should('eq', 200);
}

```

Screen Shot 2021-07-20 at 11 30 46 PM

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts
Comment (bgambocjaviniar) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/regression/quick-order/b2b-quick-order.e2e-spec.ts]: Have you tried it? 

What you have inside this `before` removes all session, cookies, storage, etc
Basically, it resets the app to be like in a vanilla state.

You can verify it gets trigged if you put a log to see it firing for each of the `it`.

Also, don't forget to try it out as this solution is better than clearing the list for each new `it`. We want the test to not be dependent on each other as what is happening right now.

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts
Comment (bgambocjaviniar) [reply projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-quick-order.ts:89]:
```suggestion
  cy.get(`.quick-order-footer .clear-button`).click();
  alerts.getSuccessAlert().should('contain', `Quick order list has been cleared`);
```

from import * as alerts from helpers/global-message

File: projects/storefrontapp-e2e-cypress/cypress.json
Comment (bgambocjaviniar) [reply projects/storefrontapp-e2e-cypress/cypress.json:12]: just a reminder to NOT forget to remove before merging these config changes to develop

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]
Almost there~

**All things are fine in here** . Again just don't forget to update all these configs for the web server api as the main CI server does not have your latest sampled ata.

BUT fix this before I approve https://github.com/SAP/spartacus/issues/13232
and add a test case afterwards in this PR

File: .env-cmdrc
Comment (psmul) [inline .env-cmdrc:3]: Please remember to revert api url changes PR-wide

**Review by psmul — APPROVED** [review]

**Pull request merged**
